src/gateways/stdioToWs.ts

```diff
@@ -0,0 +1,158 @@
+import express from 'express'
+import cors from 'cors'
+import { spawn, ChildProcessWithoutNullStreams } from 'child_process'
+import { Server } from '@modelcontextprotocol/sdk/server/index.js'
+import { JSONRPCMessage } from '@modelcontextprotocol/sdk/types.js'
+import { Logger } from '../types.js'
+import { getVersion } from '../lib/getVersion.js'
+import { WebSocketServerTransport } from '../server/websocket.js'
+import { onSignals } from '../lib/onSignals.js'
+
+export interface StdioToWsArgs {
+  stdioCmd: string
+  port: number
+  baseUrl: string
+  messagePath: string
+  logger: Logger
+  enableCors: boolean
+  healthEndpoints: string[]
+}
+
+export async function stdioToWs(args: StdioToWsArgs) {
+  const {
+    stdioCmd,
+    port,
+    baseUrl,
+    messagePath,
+    logger,
+    healthEndpoints,
+    enableCors,
+  } = args
+  const hostname = baseUrl ? new URL(baseUrl).hostname : '0.0.0.0'
+
+  logger.info(`  - port: ${port}`)
+  logger.info(`  - stdio: ${stdioCmd}`)
+  if (baseUrl) {
+    logger.info(`  - baseUrl: ${baseUrl}`)
+  }
+  logger.info(`  - messagePath: ${messagePath}`)
+  logger.info(`  - CORS enabled: ${enableCors}`)
+  logger.info(
+    `  - Health endpoints: ${healthEndpoints.length ? healthEndpoints.join(', ') : '(none)'}`,
+  )
+
+  let wsTransport: WebSocketServerTransport | null = null
+  let child: ChildProcessWithoutNullStreams | null = null
+  let isReady = false
+
+  const cleanup = () => {
+    if (wsTransport) {
+      wsTransport.close().catch((err) => {
+        logger.error(`Error stopping WebSocket server: ${err.message}`)
+      })
+    }
+    if (child) {
+      child.kill()
+    }
+  }
+
+  onSignals({
+    logger,
+    cleanup,
+  })
+
+  if (healthEndpoints.length > 0) {
+    const app = express()
+    if (enableCors) {
+      app.use(cors())
+    }
+    for (const ep of healthEndpoints) {
+      app.get(ep, (_req: express.Request, res: express.Response) => {
+        if (child?.killed) {
+          res.status(500).send('Child process has been killed')
+        }
+        if (!isReady) {
+          res.status(500).send('Server is not ready')
+        } else {
+          res.send('ok')
+        }
+      })
+    }
+    app.listen(port, hostname, () => {
```

Comment:
It's not possible to use the same port since it fails when the transport starts because port is already taken.

Reply:
Gotcha, should have tested this more 😅

Reply:
No worries! Can't wait to see it merged!